pandas/_libs/index.pyi

```diff
@@ -54,6 +54,7 @@ class UInt16Engine(IndexEngine): ...
 class UInt8Engine(IndexEngine): ...
 class ObjectEngine(IndexEngine): ...
 class StringEngine(IndexEngine): ...
+class StringObjectEngine(ObjectEngine): ...
```

Comment:
Hmm would it be better to call this `StrEngine`? Or where does the term `StringObject` come from?

Reply:
It was meant to be read as "string-objectengine", i.e. essentially just the object engine, but we know that we only use it for strings (and so the `_check_type` can be specialized).

But I don't mind the name exactly (although `StrEngine` might also be confusing, because we currently use this for both str and string dtypes)